python-code/code.py

```diff
@@ -0,0 +1,125 @@
+    "Takes in the full url and returns the full file path"
+    "File names are ZIP_COUNTY_032010.xlsx"
+    curpath = os.path.abspath(os.curdir) #get current working directory
+    full_path = ''
+    if csv_file:
+        #If we are passing in a csv, change xlsx to .csv
+        csv_file_name = url.split("\\")[-1][:-5] + ".csv"
+        full_path = os.path.join(curpath, "download\csv", csv_file_name)
+    else:
+        url_name = url.split('/')[-1] # gets file name
+        #switching file names to be YYYY-MM for better file management
+        url_file_name = generate_file_name_from_url(url) + ".xlsx"
+        full_path = os.path.join(curpath, "download\excel", url_file_name)
+    return full_path
+
+def download_file(url):
+    #With a full url, downloads the full file in chunks.
+    #Able to handle large files.
+    full_file_path = get_file_path(url)
+    r = requests.get(url)
+    with open(full_file_path, 'wb') as f:
+        for chunk in r.iter_content(chunk_size=1024):
+            if chunk: # filter out keep-alive new chunks
+                f.write(chunk)
+    return full_file_path
+
+"""
+Census State & County FIPS Data
+HUD uses ZIP & FIPS data. We need to grab the FIPS to county
+name data to be able to merge and create the cross lookup
+"""
+
+census_fips_url = "https://www2.census.gov/geo/docs/reference/codes/files/national_county.txt"
+#the FIPS data does not come with column names
+census_col_names = ["STATE","STATEFP","COUNTYFP","COUNTYNAME","CLASSFP"]
+
+# Open url, read in the data with the column names, and convert specific columns to str.
+# When Pandas reads these columns, it automatilcally intrepets them as INTS
+fips_df = pd.read_table(
+    census_fips_url,
+    sep=",",
+    names=census_col_names,
+    converters={'STATEFP': str,'COUNTYFP': str,'CLASSFP': str}
+)
+
+# Combine State & County FP to generate the full FIPS code for easier lookup
+fips_df["STCOUNTYFP"] = fips_df["STATEFP"] + fips_df["COUNTYFP"]
+#Dropping STATFP & COUNTYFP as we no longer need them
+fips_df = fips_df[["STCOUNTYFP", "STATE" ,"COUNTYNAME", "CLASSFP"]]
+
+# Get current year to handle future runs of this file
+now = dt.datetime.now()
+cur_year = now.year
+
+def get_files_url(month, year):
+    monthyear = month + str(year)
+    return "https://www.huduser.gov/portal/datasets/usps/ZIP_COUNTY_{}.xlsx".format(monthyear)
+
+"""
+Main loop
+"""
+def main()
+    # from the beginning of hud year data to current year
+    for year in range(2010, cur_year+1):
+        #hud files are based on quarters
+        for month in ["03", "06", "09", "12"]:
+            #generate the HUDs url
+            url = get_files_url(month, year)
+            #download the file
+            full_file_path = download_file(url)
+            #open and get the excel dataframe
+            excel_df = process_excel_file(full_file_path)
+            #merge the excel file with the fips data
+            merged_df = fips_df.merge(excel_df)
```

Comment:
It would be better if this function did not refer to `fits_df`, which is defined outside of the function. That should be passed in as an argument or defined within the function.